pubsub-open-telemetry/publisher.cc

```diff
@@ -0,0 +1,89 @@
+#include "google/cloud/opentelemetry/configure_basic_tracing.h"
+#include "google/cloud/opentelemetry/trace_exporter.h"
+#include "google/cloud/opentelemetry_options.h"
+#include "parse_args.h"
+#include <opentelemetry/sdk/trace/batch_span_processor_factory.h>
+#include <opentelemetry/sdk/trace/batch_span_processor_options.h>
+#include <opentelemetry/sdk/trace/processor.h>
+#include <opentelemetry/sdk/trace/tracer_provider_factory.h>
+#include <opentelemetry/trace/provider.h>
+#include <iostream>
+#include <string_view>
+
+// Create a few namespace aliases to make the code easier to read.
+namespace gc = ::google::cloud;
+namespace pubsub = gc::pubsub;
+namespace otel = gc::otel;
+
+std::string GeneratePayload(int payload_size) {
+  auto gen = google::cloud::internal::DefaultPRNG(std::random_device{}());
+  const std::string charset = "abcdefghijklmnopqrstuvwxyz";
+  std::uniform_int_distribution<std::size_t> rd(0, charset.size() - 1);
+
+  std::string result(payload_size, '0');
+  std::generate(result.begin(), result.end(),
+                [&rd, &gen, &charset]() { return charset[rd(gen)]; });
+  return result;
+}
+
+int main(int argc, char* argv[]) try {
+  auto args = ParseArguments(argc, argv);
+  if (args.project_id.empty() && args.topic_id.empty()) {
+    return 1;
+  }
+
+  std::cout << "Using project `" << args.project_id << "` and topic `"
+            << args.topic_id << "`\n";
+  auto exporter = otel::MakeTraceExporter(gc::Project(args.project_id));
+  opentelemetry::sdk::trace::BatchSpanProcessorOptions span_options;
+  span_options.max_queue_size = args.max_queue_size;
+  auto processor = opentelemetry::sdk::trace::BatchSpanProcessorFactory::Create(
+      std::move(exporter), span_options);
+  auto provider = opentelemetry::sdk::trace::TracerProviderFactory::Create(
+      std::move(processor));
+  opentelemetry::trace::Provider::SetTracerProvider(std::move(provider));
+
+  auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(
+      pubsub::Topic(args.project_id, args.topic_id), args.publisher_options));
+
+  std::cout << "Publishing " << std::to_string(args.message_count)
+            << " message(s) with payload size "
+            << std::to_string(args.message_size) << "...\n";
+  std::vector<gc::future<void>> ids;
+  for (int i = 0; i < args.message_count; i++) {
+    auto id = publisher
+                  .Publish(pubsub::MessageBuilder()
+                               .SetData(GeneratePayload(args.message_size))
+                               .Build())
+                  .then([](gc::future<gc::StatusOr<std::string>> f) {
+                    auto id = f.get();
+                    if (!id) {
+                      std::cout << "Error in publish: " << id.status() << "\n";
+                      return;
+                    }
+                    std::cout << "Sent message with id: (" << *id << ")\n";
+                  });
+    ids.push_back(std::move(id));
+  }
+  for (auto& id : ids) id.get();
```

Comment:
Since exceptions are required, we might as well take advantage of them:

```suggestion
  std::vector<gc::future<std::string>> ids;
  for (int i = 0; i < args.message_count; i++) {
    auto id = publisher
                  .Publish(pubsub::MessageBuilder()
                               .SetData(GeneratePayload(args.message_size))
                               .Build())
                  .then([](gc::future<gc::StatusOr<std::string>> f) {
                    return f.get().value();
                  });
    ids.push_back(std::move(id));
  }
  for (auto& id : ids) try {
      std::cout << "Sent message with id: " << id.get() << "\n";
   } catch(std::exception const& ex) {
     std::cout << "Error in publish: " << ex.what() << "\n";
   } 
```